application.go

```diff
@@ -104,6 +77,92 @@ func IsApplicationEqual(a, b Application) bool {
 		a.Handlers().IsEqual(b.Handlers())
 }
 
+// ForeignMessageNames returns the subset of message names used by an
+// application that must be communicated beyond the scope of that application.
+//
+// This includes:
+//  - commands that are produced by the application, but consumed elsewhere
+//  - commands that are consumed by the application, but produced elsewhere
+//  - events that are consumed by the application, but produced elsewhere
+func ForeignMessageNames(app Application) EntityMessageNames {
+	m := app.MessageNames()
+	f := EntityMessageNames{
+		Roles:    message.NameRoles{},
+		Produced: message.NameRoles{},
+		Consumed: message.NameRoles{},
+	}
+
+	for mt, r := range m.Produced {
+		if m.Consumed.Has(mt) {
+			continue
+		}
+
+		// Commands MUST always have a handler. Therefore, any command that is
+		// produced by this application, but not consumed by this application is
+		// considered foreign.
+		if r == message.CommandRole {
+			f.Roles.Add(mt, r)
+			f.Produced.Add(mt, r)
+		}
+	}
+
+	for mt, r := range m.Consumed {
+		if m.Produced.Has(mt) {
+			continue
+		}
+
+		// Any message type is considered foreign if it needs to be obtained from
```

Comment:
Are these actually "message type"'s as the comment mentions?

Reply:
My terminology is possibly a bit lax, but yes -- the message "name" refers to the name of the type. I've reworded it anyway.